### PR TITLE
feat(expect): add `toHaveBeenCalledAfter` and `toHaveBeenCalledBefore` utility

### LINKED ISSUE #6045: Integrate `toHaveBeenCalledBefore` and `toHaveBeenCalledAfter` to core
### Clear and concise description of the problem

Hello,

Knowing when a mock is called relatively to another is very useful to ensure implementation order is correct. Actually, these is no expect function to know that. Do you thing it could be possible to add some or have you any recommanded way to check that a mock `vi.fn` is called before or after another one.

### Suggested solution

Integrate https://github.com/jest-community/jest-extended/blob/main/src/matchers/toHaveBeenCalledBefore.js to core of vitest.

### Alternative

_No response_

### Additional context

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by Barbapapazes**
### Description

fix #6045

Hello 👋,

This is my first contribution to Vitest so I really appreciate your feedback.

This PR adds two new test APIs:

- `toHaveBeenCalledBefore` - checks if a spy was called before another spy
- `toHaveBeenCalledAfter` - checks if a spy was called after another spy

The implementation will take a look at the invocation order (first call) of the spies and compare them to determine if the spies were called in the correct order.

I add them to `packages/expect/src/jest-expect.ts` but I'm not sure if this is the best place for them. I also update types but I'm not sure if I did it correctly so I would appreciate your feedback on that.

I add related tests to for these new APIs.

Once everything is ok for you, I will add related documentation!

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: packages/expect/src/jest-expect.ts [hunk 1]
```diff
@@ -699,6 +699,42 @@ export const JestChaiExpect: ChaiPlugin = (chai, utils) => {
       )
     },
   )
+  def(
+    ['toHaveBeenCalledBefore'],
+    function (resultSpy: MockInstance) {
+      const expectSpy = getSpy(this)
+
+      const [firstExpectSpyCall] = expectSpy.mock.invocationCallOrder
```
Comment (sheremet-va) [inline packages/expect/src/jest-expect.ts:707]: Why first? Shouldn't it be the smallest?

https://github.com/jest-community/jest-extended/blob/main/src/matchers/toHaveBeenCalledAfter.js#L43

File: packages/expect/src/jest-expect.ts [hunk 2]
```diff
@@ -699,6 +699,42 @@ export const JestChaiExpect: ChaiPlugin = (chai, utils) => {
       )
     },
   )
+  def(
+    ['toHaveBeenCalledBefore'],
+    function (resultSpy: MockInstance) {
+      const expectSpy = getSpy(this)
```
Comment (sheremet-va) [inline packages/expect/src/jest-expect.ts:705]: We also need to validate the `resultSpy` to be a mock function

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: packages/expect/src/jest-expect.ts (hunk 2 above)
Comment (Barbapapazes) [reply packages/expect/src/jest-expect.ts:705]: Indeed, I will update that part.

File: packages/expect/src/jest-expect.ts (hunk 1 above)
Comment (Barbapapazes) [reply packages/expect/src/jest-expect.ts:707]: As the calls are ordered in the table, I thought that taking the first was enough, but maybe I'm wrong?

File: packages/expect/src/jest-expect.ts [hunk 3]
```diff
@@ -699,6 +699,42 @@ export const JestChaiExpect: ChaiPlugin = (chai, utils) => {
       )
     },
   )
+  def(
+    ['toHaveBeenCalledBefore'],
+    function (resultSpy: MockInstance) {
```
[the remark this answers is not in the document]
Comment (Barbapapazes) [reply packages/expect/src/jest-expect.ts:704]: yes, this is definitely an interesting option.

File: packages/expect/src/jest-expect.ts [hunk 4]
```diff
@@ -699,6 +699,82 @@ export const JestChaiExpect: ChaiPlugin = (chai, utils) => {
...
+    function (resultSpy: MockInstance, failIfNoFirstInvocation = true) {
+      const expectSpy = getSpy(this)
+
+      if (!isMockFunction(resultSpy)) {
+        throw new TypeError(
+          `${utils.inspect(resultSpy)} is not a spy or a call to a spy`,
+        )
+      }
+
+      if (!isMockFunction(expectSpy)) {
```
Comment (sheremet-va) [inline packages/expect/src/jest-expect.ts:729]: `getSpy` already checks if it's a spy

File: packages/expect/src/jest-expect.ts [hunk 5]
```diff
@@ -699,6 +699,82 @@ export const JestChaiExpect: ChaiPlugin = (chai, utils) => {
       )
     },
   )
+
+  /**
+   * Used for `toHaveBeenCalledBefore` and `toHaveBeenCalledAfter` to determine if the expected spy was called before the result spy.
+   */
+  function predicate(expectInvocationCallOrder: number[], resultInvocationCallOrder: number[], failIfNoFirstInvocation: number): boolean {
```
Comment (sheremet-va) [inline packages/expect/src/jest-expect.ts:706]: I think the name is too generic for this giant file

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: packages/expect/src/jest-expect.ts (hunk 5 above)
Comment (Barbapapazes) [reply packages/expect/src/jest-expect.ts:706]: Indeed but should I find a better name or is there a file where I could put this function?

Comment (sheremet-va) [reply packages/expect/src/jest-expect.ts:706]: Let's find a better name. The refactoring should be done separetly

File: packages/expect/src/jest-expect.ts (hunk 4 above)
Comment (Barbapapazes) [reply packages/expect/src/jest-expect.ts]: Alright, I'll remove the check

**Review by sheremet-va — APPROVED** [review]
Sorry, this got lost in notifications. Looks good to me!

**Pull request merged**